python/cudf/cudf/tests/dataset_generator.py

```diff
@@ -0,0 +1,145 @@
+        self,
+        cardinality=100,
+        null_frequency=0.1,
+        ty=lambda g: g.address.country,
+        is_sorted=True,
+    ):
+        self.cardinality = cardinality
+        self.null_frequency = null_frequency
+        self.ty = ty
+        self.is_sorted = is_sorted
+
+
+class Parameters:
+    """Parameters for random dataset generation
+
+    Attributes
+    ---
+    num_rows : int
+        Number of rows to generate
+    column_params : List[ColumnParams]
+        ColumnParams for each column
+    seed : int or None, default None
+        Seed for random data generation
+    """
+
+    def __init__(
+        self, num_rows=2048, column_parameters=[], seed=None,
+    ):
+        self.num_rows = num_rows
+        self.column_parameters = column_parameters
+        self.seed = seed
+
+
+SIMPLE = Parameters(
+    num_rows=2048,
+    column_parameters=[
+        ColumnParameters(
+            cardinality=100,
+            null_frequency=0.05,
+            ty=lambda g: g.address.country,
+            is_sorted=False,
+        ),
+        ColumnParameters(
+            cardinality=40,
+            null_frequency=0.2,
+            ty=lambda g: g.person.age,
+            is_sorted=True,
+        ),
+        ColumnParameters(
+            cardinality=30,
+            null_frequency=0.1,
+            ty=lambda g: g.text.color,
+            is_sorted=False,
+        ),
+        ColumnParameters(
+            cardinality=10,
+            null_frequency=0.1,
+            ty=lambda g: g.hardware.manufacturer,
+            is_sorted=False,
+        ),
+    ],
+    seed=0,
+)
+
+
+def generate(
+    path, parameters, format={"name": "parquet", "row_group_size": 64}
+):
+    """
+    Generate dataset using given parameters and write to given format
+
+    Parameters
+    ----------
+    path : str or file-like object
+        Path to write to
+    parameters : Parameters
+        Parameters specifying how to randomly generate data
+    format : Dict
+        Format to write
+    """
+
+    # Initialize seeds
+    g = Generic("es", seed=parameters.seed)
+    if parameters.seed is not None:
+        np.random.seed(parameters.seed)
+
+    # Generate data for each column
+    data = {}
+    columns_to_sort = []
+    for i, column_params in enumerate(parameters.column_parameters):
+        vals = np.array(
+            [column_params.ty(g)() for _ in range(column_params.cardinality)]
+        )
+
+        # Generate data for current column
+        data[str(i)] = np.array(
+            [
+                None
+                if np.random.rand() < column_params.null_frequency
+                else np.random.choice(vals)
+                for _ in range(parameters.num_rows)
+            ]
+        )
+
+        # Check if marked for sorting
+        if column_params.is_sorted:
+            columns_to_sort.append(str(i))
+
+    # Create DataFrame and sort columns appropriately
+    df = pd.DataFrame(data)
+    df = df.sort_values(columns_to_sort)
```

Comment:
Would also ideally use a PyArrow Table here, but that prevents us from sorting AFAIK

Reply:
Being able to generate sorted data is important so I am using `pyarrow.Array` in the 2 instances you commented on but am converting to `pandas.DataFrame` here which should be relatively fast.

Reply:
My fear is that if we have nulls we can cause implicit conversion from int --> float in going from pyarrow to pandas. Just something to keep in mind.